File: tests/cases/python/hop-says-what-it-is/pipeline.py
"""A chain whose hops are four different relations, so a chain that prints them all the same
is visibly wrong: a decorator application, a construction, an ordinary call, and a closure
written inside a body (which is not a call at all)."""


def audited(fn):
    def wrapper(*a, **kw):
        return fn(*a, **kw)
    return wrapper


class Store:
    def __init__(self, name):
        self.name = name
        self.rows = []

    def put(self, row):
        self.rows.append(row)


class Service:
    def __init__(self):
        self.store = Store("main")

    @audited
    def add(self, row):
        self.store.put(row)


def run():
    svc = Service()
    svc.add("a")
    return svc

File: tests/cases/java/hop-says-what-it-is/src/Pipeline.java
// A chain whose hops are four different relations. A chain that prints them all the same is
// visibly wrong: a construction, a constructor chaining to another constructor, a method
// reference handed over rather than called, and an ordinary invocation.
import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

class Store {
    private final List<String> rows = new ArrayList<>();
    private final String name;

    Store() {
        this("main");
    }

    Store(String name) {
        this.name = name;
    }

    void put(String row) {
        rows.add(row);
    }
}

class Service {
    private final Store store = new Store();

    void add(String row) {
        store.put(row);
    }

    Consumer<String> adder() {
        return this::add;
    }
}

public class Pipeline {
    public static void main(String[] args) {
        Service svc = new Service();
        svc.add("a");
        svc.adder().accept("b");
    }
}
